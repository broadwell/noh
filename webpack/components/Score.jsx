import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import MeasureLabelContainer from "./MeasureLabelContainer";
import CellBeat from "./CellBeat";
import CellPercussion from "./CellPercussion";
import NohkanLine from "./NohkanLine";
import DanceLine from "./DanceLine";
import ScoreTextLine from "./ScoreTextLine";

function createBeatsArray(grid) {
  const beatsArray = [];
  for (let i = 0; i < grid.length; i += 1) {
    beatsArray[grid[i].start] = grid[i].text;
  }
  return [...Array(beatsArray.length).keys()].map(i => beatsArray[i] || "");
}

class Score extends Component {
  constructor(props) {
    super(props);
    this.state = {
      previousPhrase: null,
      currentPhrase: this.props.phrases[0],
      nextPhrase: this.props.phrases[1]
    };
  }

  componentWillReceiveProps() {
    const currentPhraseId = this.determineCurrentPhrase();
    const previousPhraseId = Math.max(currentPhraseId - 1, 0);
    const nextPhraseId = Math.min(
      currentPhraseId + 1,
      this.props.phrases.length
    );
    this.setState({
      previousPhrase: this.props.phrases[previousPhraseId],
      currentPhrase: this.props.phrases[currentPhraseId],
      nextPhrase: this.props.phrases[nextPhraseId]
    });
  }

  determineCurrentPhrase() {
    if (this.props.currentTime > 0) {
      return (
        this.props.phrases.length -
        1 -
        this.props.phrases
          .filter(Boolean)
          .reverse()
          .findIndex(phrase => this.props.currentTime >= phrase.startTime.value)
      );
    }
    return 0;
  }

  render() {
    console.log(this.props.phrases[0]);
    console.log(this.props.phrases[1]);

    const prevBeatNums = this.state.previousPhrase
      ? createBeatsArray(this.state.previousPhrase.beat.grid)
      : [];
    const currentBeatNums = this.state.currentPhrase
      ? createBeatsArray(this.state.currentPhrase.beat.grid)
      : [];
    const nextBeatNums = this.state.nextPhrase
      ? createBeatsArray(this.state.nextPhrase.beat.grid)
      : [];

    const prevBeats = prevBeatNums.map((num, idx) => (
      <CellBeat beatText={num} key={`beatNum${idx}`} /> // eslint-disable-line react/no-array-index-key
    ));
    const currentBeats = currentBeatNums.map((num, idx) => (
      <CellBeat beatText={num} key={`beatNum${idx}`} /> // eslint-disable-line react/no-array-index-key
    ));
    const nextBeats = nextBeatNums.map((num, idx) => (
      <CellBeat beatText={num} key={`beatNum${idx}`} /> // eslint-disable-line react/no-array-index-key
    ));
    const prevPercussion = this.state.previousPhrase
      ? this.state.previousPhrase.percussion.value
      : "None";
    const currentPercussion = this.state.currentPhrase
      ? this.state.currentPhrase.percussion.value
      : "None";
    const nextPercussion = this.state.nextPhrase
      ? this.state.nextPhrase.percussion.value
      : "None";
    return (
      <div className="score">
        <div className="measure measure--prev">
          <MeasureLabelContainer previous />
          <div className="measure__grid-container">
            <div className="measure__channel">{prevBeats}</div>
            <div className="measure__channel measure__channel--large">
              <ScoreTextLine
                grid={
                  this.state.previousPhrase
                    ? this.state.previousPhrase.syllableText.grid
                    : []
                }
                length={prevBeatNums.length}
              />
            </div>
            <div className="measure__channel">
              <CellPercussion
                text={prevPercussion}
                length={prevBeatNums.length}
              />
            </div>
            <div className="measure__channel">
              <NohkanLine
                grid={
                  this.state.previousPhrase
                    ? this.state.previousPhrase.nohkan.grid
                    : []
                }
                length={prevBeatNums.length}
              />
            </div>
            <div className="measure__channel">
              <DanceLine
                grid={
                  this.state.previousPhrase
                    ? this.state.previousPhrase.dance.grid
                    : []
                }
                length={prevBeatNums.length}
              />
            </div>
          </div>
        </div>
        <div className="measure measure--current">
          <MeasureLabelContainer />
          <div className="measure__grid-container">
            <div className="measure__channel">{currentBeats}</div>
            <div className="measure__channel measure__channel--large">
              <ScoreTextLine
                grid={
                  this.state.currentPhrase
                    ? this.state.currentPhrase.syllableText.grid
                    : []
                }
                length={currentBeatNums.length}
              />
            </div>
            <div className="measure__channel">
              <CellPercussion
                text={currentPercussion}
                length={currentBeatNums.length}
              />
            </div>
            <div className="measure__channel">
              <NohkanLine
                grid={
                  this.state.currentPhrase
                    ? this.state.currentPhrase.nohkan.grid
                    : []
                }
                length={currentBeatNums.length}
              />
            </div>
            <div className="measure__channel">
              <DanceLine
                grid={
                  this.state.currentPhrase
                    ? this.state.currentPhrase.dance.grid
                    : []
                }
                length={currentBeatNums.length}
              />
            </div>
          </div>
        </div>
        <div className="measure measure--next">
          <MeasureLabelContainer next />
          <div className="measure__grid-container">
            <div className="measure__channel">{nextBeats}</div>
            <div className="measure__channel measure__channel--large">
              <ScoreTextLine
                grid={
                  this.state.nextPhrase
                    ? this.state.nextPhrase.syllableText.grid
                    : []
                }
                length={nextBeatNums.length}
              />
            </div>
            <div className="measure__channel">
              <CellPercussion
                text={nextPercussion}
                length={nextBeatNums.length}
              />
            </div>
            <div className="measure__channel">
              <NohkanLine
                grid={
                  this.state.nextPhrase ? this.state.nextPhrase.nohkan.grid : []
                }
                length={nextBeatNums.length}
              />
            </div>
            <div className="measure__channel">
              <DanceLine
                grid={
                  this.state.nextPhrase ? this.state.nextPhrase.dance.grid : []
                }
                length={nextBeatNums.length}
              />
            </div>
          </div>
        </div>
      </div>
    );
  }
}

Score.propTypes = {
  currentTime: PropTypes.number.isRequired,
  phrases: PropTypes.arrayOf(
    PropTypes.shape({
      startTime: PropTypes.shape({}),
      beat: PropTypes.shape({
        grid: PropTypes.arrayOf(PropTypes.shape({}))
      }),
      dance: PropTypes.shape({}),
      nohkan: PropTypes.shape({}),
      percussion: PropTypes.shape({ value: PropTypes.string.isRequired }),
      phrase: PropTypes.string,
      syllableNumber: PropTypes.shape({}),
      syllableText: PropTypes.shape({}),
      text: PropTypes.shape({}),
      vocalRange: PropTypes.shape({})
    })
  ).isRequired
};

const mapStateToProps = state => ({
  currentTime: state.currentTime
});

export const Unwrapped = Score;
export default connect(mapStateToProps)(Score);
