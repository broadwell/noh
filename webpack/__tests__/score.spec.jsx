import React from "react";
import { shallow } from "enzyme";
import phrases from "./__fixtures__/phrases.json";
import Score from "../components/Score";

describe("<Score>", () => {
  it("renders as expected", () => {
    const component = shallow(<Score phrases={phrases.phrases} />);
    expect(component).toMatchSnapshot();
  });
});
