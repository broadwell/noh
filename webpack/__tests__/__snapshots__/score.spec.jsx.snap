// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Score> renders as expected 1`] = `
<div
  className="score"
>
  <div
    className="measure measure--prev"
  >
    <MeasureLabelContainer
      next={false}
      previous={true}
    />
    <div
      className="measure__grid-container"
    >
      <div
        className="measure__channel"
      >
        <CellBeat
          beatText=""
          key="beatNum0"
        />
        <CellBeat
          beatText="1"
          key="beatNum1"
        />
        <CellBeat
          beatText=""
          key="beatNum2"
        />
        <CellBeat
          beatText="2"
          key="beatNum3"
        />
        <CellBeat
          beatText=""
          key="beatNum4"
        />
        <CellBeat
          beatText="3"
          key="beatNum5"
        />
        <CellBeat
          beatText="4"
          key="beatNum6"
        />
        <CellBeat
          beatText=""
          key="beatNum7"
        />
        <CellBeat
          beatText="5"
          key="beatNum8"
        />
        <CellBeat
          beatText="6"
          key="beatNum9"
        />
        <CellBeat
          beatText=""
          key="beatNum10"
        />
        <CellBeat
          beatText="7"
          key="beatNum11"
        />
        <CellBeat
          beatText=""
          key="beatNum12"
        />
        <CellBeat
          beatText="8"
          key="beatNum13"
        />
      </div>
      <div
        className="measure__channel measure__channel--large"
      >
        <div
          className="cell cell--1 cell--text"
        />
        <div
          className="cell cell--1 cell--text"
        />
        <div
          className="cell cell--1 cell--text"
        />
        <div
          className="cell cell--1 cell--text"
        />
        <div
          className="cell cell--1 cell--text"
        >
          ke
        </div>
        <div
          className="cell cell--1 cell--text"
        >
          i
        </div>
        <div
          className="cell cell--1 cell--text"
        >
          ko
        </div>
        <div
          className="cell cell--1 cell--text"
        >
          o
        </div>
        <div
          className="cell cell--1 cell--text"
        >
          te
        </div>
        <div
          className="cell cell--1 cell--text"
        >
          n
        </div>
        <div
          className="cell cell--1 cell--text"
        >
          no
        </div>
        <div
          className="cell cell--1 cell--text"
        >
          o
        </div>
        <div
          className="cell cell--1 cell--text"
        />
        <div
          className="cell cell--1 cell--text"
        />
      </div>
      <div
        className="measure__channel"
      >
        <div
          className="cell cell--14"
        >
          Tsuzuke
        </div>
      </div>
      <div
        className="measure__channel"
      >
        <div
          className="cell cell--2 cell--dance"
        />
        <div
          className="cell cell--9 cell--dance"
        >
          grid
        </div>
        <div
          className="cell cell--3 cell--dance"
        />
      </div>
      <div
        className="measure__channel"
      >
        <div
          className="cell cell--3"
        />
        <div
          className="cell cell--6 cell--dance"
        >
          Sits back on his knee (9)
        </div>
        <div
          className="cell cell--5"
        />
      </div>
    </div>
  </div>
  <div
    className="measure measure--current"
  >
    <MeasureLabelContainer
      next={false}
      previous={false}
    />
    <div
      className="measure__grid-container"
    >
      <div
        className="measure__channel"
      >
        <CellBeat
          beatText=""
          key="beatNum0"
        />
        <CellBeat
          beatText="1"
          key="beatNum1"
        />
        <CellBeat
          beatText=""
          key="beatNum2"
        />
        <CellBeat
          beatText="2"
          key="beatNum3"
        />
        <CellBeat
          beatText=""
          key="beatNum4"
        />
        <CellBeat
          beatText="3"
          key="beatNum5"
        />
        <CellBeat
          beatText=""
          key="beatNum6"
        />
        <CellBeat
          beatText="4"
          key="beatNum7"
        />
        <CellBeat
          beatText=""
          key="beatNum8"
        />
        <CellBeat
          beatText="5"
          key="beatNum9"
        />
        <CellBeat
          beatText=""
          key="beatNum10"
        />
        <CellBeat
          beatText="6"
          key="beatNum11"
        />
        <CellBeat
          beatText=""
          key="beatNum12"
        />
        <CellBeat
          beatText="7"
          key="beatNum13"
        />
        <CellBeat
          beatText=""
          key="beatNum14"
        />
        <CellBeat
          beatText="8"
          key="beatNum15"
        />
      </div>
      <div
        className="measure__channel measure__channel--large"
      >
        <div
          className="cell cell--2 cell--text"
        />
        <div
          className="cell cell--5 cell--text"
        >
          Tachibana no michinari to wa,
          <span
            className="cell__character cell__character--begin cell__character--jiutai"
          />
          <span
            className="cell__range--low"
          />
        </div>
        <div
          className="cell cell--1 cell--text"
        >
          waga
          <span
            className="cell__character cell__character--end cell__character--jiutai"
          />
          <span
            className="cell__range--medium"
          />
        </div>
        <div
          className="cell cell--2 cell--text"
        >
          koto nari.
          <span
            className="cell__character cell__character--begin cell__character--end cell__character--shite"
          />
          <span
            className="cell__range--high"
          />
        </div>
        <div
          className="cell cell--6 cell--text"
        />
      </div>
      <div
        className="measure__channel"
      >
        <div
          className="cell cell--2"
        />
        <div
          className="cell cell--12"
        >
          Tsuzuke
        </div>
        <div
          className="cell cell--2"
        />
      </div>
      <div
        className="measure__channel"
      >
        <div
          className="cell cell--2 cell--nohkan"
        >
          grid
        </div>
        <div
          className="cell cell--8 cell--nohkan"
        >
          grid
        </div>
        <div
          className="cell cell--6 cell--nohkan"
        >
          grid
        </div>
      </div>
      <div
        className="measure__channel"
      >
        <div
          className="cell cell--3"
        />
        <div
          className="cell cell--6 cell--dance"
        >
          Sits back on his knee (9)
        </div>
        <div
          className="cell cell--7"
        />
      </div>
    </div>
  </div>
  <div
    className="measure measure--next"
  >
    <MeasureLabelContainer
      next={true}
      previous={false}
    />
    <div
      className="measure__grid-container"
    >
      <div
        className="measure__channel"
      >
        <CellBeat
          beatText=""
          key="beatNum0"
        />
        <CellBeat
          beatText="1"
          key="beatNum1"
        />
        <CellBeat
          beatText=""
          key="beatNum2"
        />
        <CellBeat
          beatText="2"
          key="beatNum3"
        />
        <CellBeat
          beatText=""
          key="beatNum4"
        />
        <CellBeat
          beatText="3"
          key="beatNum5"
        />
        <CellBeat
          beatText="4"
          key="beatNum6"
        />
        <CellBeat
          beatText=""
          key="beatNum7"
        />
        <CellBeat
          beatText="5"
          key="beatNum8"
        />
        <CellBeat
          beatText="6"
          key="beatNum9"
        />
        <CellBeat
          beatText=""
          key="beatNum10"
        />
        <CellBeat
          beatText="7"
          key="beatNum11"
        />
        <CellBeat
          beatText=""
          key="beatNum12"
        />
        <CellBeat
          beatText="8"
          key="beatNum13"
        />
      </div>
      <div
        className="measure__channel measure__channel--large"
      >
        <div
          className="cell cell--1 cell--text"
        />
        <div
          className="cell cell--1 cell--text"
        />
        <div
          className="cell cell--1 cell--text"
        />
        <div
          className="cell cell--1 cell--text"
        />
        <div
          className="cell cell--1 cell--text"
        >
          ke
        </div>
        <div
          className="cell cell--1 cell--text"
        >
          i
        </div>
        <div
          className="cell cell--1 cell--text"
        >
          ko
        </div>
        <div
          className="cell cell--1 cell--text"
        >
          o
        </div>
        <div
          className="cell cell--1 cell--text"
        >
          te
        </div>
        <div
          className="cell cell--1 cell--text"
        >
          n
        </div>
        <div
          className="cell cell--1 cell--text"
        >
          no
        </div>
        <div
          className="cell cell--1 cell--text"
        >
          o
        </div>
        <div
          className="cell cell--1 cell--text"
        />
        <div
          className="cell cell--1 cell--text"
        />
      </div>
      <div
        className="measure__channel"
      >
        <div
          className="cell cell--14"
        >
          Tsuzuke
        </div>
      </div>
      <div
        className="measure__channel"
      >
        <div
          className="cell cell--2 cell--dance"
        />
        <div
          className="cell cell--9 cell--dance"
        >
          grid
        </div>
        <div
          className="cell cell--3 cell--dance"
        />
      </div>
      <div
        className="measure__channel"
      >
        <div
          className="cell cell--3"
        />
        <div
          className="cell cell--6 cell--dance"
        >
          Sits back on his knee (9)
        </div>
        <div
          className="cell cell--5"
        />
      </div>
    </div>
  </div>
</div>
`;
